modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/config/Config.scala

```diff
@@ -0,0 +1,117 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.rdbloader.config
+
+import java.net.URI
+
+import scala.concurrent.duration.{Duration, FiniteDuration}
+
+import cats.effect.Sync
+import cats.data.EitherT
+import cats.syntax.either._
+
+import io.circe._
+import io.circe.generic.semiauto._
+
+import org.http4s.{ParseFailure, Uri}
+
+import com.snowplowanalytics.snowplow.rdbloader.config.Config._
+import com.snowplowanalytics.snowplow.rdbloader.common.config.{Step, ConfigUtils, Region}
+import com.snowplowanalytics.snowplow.rdbloader.common.S3
+import com.snowplowanalytics.snowplow.rdbloader.common._
+
+/**
+ * Main config file parsed from HOCON
+ * @tparam D kind of supported warehouse
+ */
+final case class Config[+D <: StorageTarget](region: Region,
+                                             jsonpaths: Option[S3.Folder],
+                                             monitoring: Monitoring,
+                                             messageQueue: String,
+                                             storage: D,
+                                             steps: Set[Step])
+
+object Config {
+
+  val MetricsDefaultPrefix = "snowplow.rdbloader"
+
+  def fromString[F[_]: Sync](s: String): EitherT[F, String, Config[StorageTarget]] =
+    fromString(s, implicits().configDecoder)
+
+  def fromString[F[_]: Sync](s: String, configDecoder: Decoder[Config[StorageTarget]]): EitherT[F, String, Config[StorageTarget]] =  {
+    implicit val implConfigDecoder: Decoder[Config[StorageTarget]] = configDecoder
+    ConfigUtils.fromStringF[F, Config[StorageTarget]](s)
+  }
+
+  final case class Monitoring(snowplow: Option[SnowplowMonitoring], sentry: Option[Sentry], metrics: Option[Metrics], webhook: Option[Webhook], folders: Option[Folders])
+  final case class SnowplowMonitoring(appId: String, collector: String)
+  final case class Sentry(dsn: URI)
+  final case class Metrics(statsd: Option[StatsD], stdout: Option[Stdout])
+  final case class StatsD(hostname: String, port: Int, tags: Map[String, String], prefix: Option[String])
+  final case class Stdout(prefix: Option[String])
+  final case class Webhook(endpoint: Uri, tags: Map[String, String])
+  final case class Folders(period: FiniteDuration, staging: S3.Folder, since: Option[FiniteDuration], shredderOutput: S3.Folder, until: Option[FiniteDuration])
+
+
+  /**
+   * All config implicits are put into case class because we want to make region decoder
+   * replaceable to write unit tests for config parsing.
+   */
+  final case class implicits(regionConfigDecoder: Decoder[Region] = Region.regionConfigDecoder) {
```

Comment:
Hey @spenes! Just noticed this one and cannot decide how I feel about it. The problem is that we have IO (`getRegion` via default chain) in something that supposed to be very pure - decoders.

Can we instead have something like:

```scala
implicit val pureConfigDecoder: Decoder[Config] = ???    // usual decoder
def getRegion: F[Option[Region]] = ???
def replaceRegion(config: Config): F[Config] = ???       // this one calls provider chain when it's necessary
def fromString: EitherT[F, String, Config] = {
  ...
  replaceRegion
  ...
}
```

The point as usual just to separate pure and impure.

Reply:
Hey @chuwy! I see your point definitely but I have some concerns about that approach. If we do that, I guess we need to make the type of the config's region field `Option[String]`. In that case, we can replace region if it is None but its type will be still `Option` and it will propagate to all the parts which uses region.

Another option is creating two different config classes with one of them has region as `Option[String]` type and other one has `Region` type. And convert from one to other right after parsing the config with replacing the region. But there will be almost same two config classes with only difference is type of the region field.

 I can't find any better solution to this problem currently. Do you think would it be okay to make an exception in here ? I know it will not be pure but at least we know if there is any failure in there, it will fail in the very beginning in the application without propagating too much.

Reply:
What about having a default `""` so that it's not an `Option` ? And you can still call `replaceRegion`

Not super pretty I know

Reply:
In that case, config parsing actually depends on `replaceRegion` but we are making it very implicit IMO. I am not sure whether we should go with that.